### PR TITLE
[SLP]Support minbitwidth analisys for buildvector nodes.

### PR DESCRIPTION
**Pull request by alexey-bataev**
Metric: size..text

Program                                                                                                                                                size..text
                                                                                                                                                       exp           ref        diff
                                                                                  test-suite :: MultiSource/Benchmarks/mediabench/gsm/toast/toast.test    42906.00    42986.00  0.2%
                                                                           test-suite :: MultiSource/Benchmarks/MiBench/telecomm-gsm/telecomm-gsm.test    42909.00    42989.00  0.2%
                                                                                   test-suite :: External/SPEC/CINT2017rate/525.x264_r/525.x264_r.test   664581.00   664661.00  0.0%
                                                                                  test-suite :: External/SPEC/CINT2017speed/625.x264_s/625.x264_s.test   664581.00   664661.00  0.0%

Less is better.

Replaces `buildvector <p x in> + trunc <p x in> to <p x im>` sequences to
`buildvector <p x im> of { trunc in to im }` scalars, which is free in
most cases, results in better code.

### CODE REVIEW

**Comment by alexey-bataev** [comment]
Ping!

**Comment by alexey-bataev** [comment]
Ping!

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp [hunk 1]
```diff
@@ -8175,6 +8178,48 @@ class BoUpSLP::ShuffleCostEstimator : public BaseShuffleAnalysis {
...
+      if (EScalarTy != ScalarTy) {
+        bool IsSigned = !isKnownNonNegative(V, SimplifyQuery(*R.DL));
+        unsigned CastOpcode = Instruction::Trunc;
+        unsigned DstSz = R.DL->getTypeSizeInBits(ScalarTy);
+        unsigned SrcSz = R.DL->getTypeSizeInBits(EScalarTy);
+        if (DstSz > SrcSz)
+          CastOpcode = IsSigned ? Instruction::SExt : Instruction::ZExt;
+        return TTI.getCastInstrCost(
+            CastOpcode, VectorType::get(ScalarTy, VecTy->getElementCount()),
+            VecTy, TTI::CastContextHint::None, TTI::TCK_RecipThroughput);
```
Comment (RKSimon) [inline llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:8219]: Should CostKind be hard coded like this?

**Comment by alexey-bataev** [comment]
Ping!

**Review by RKSimon — APPROVED** [review]
LGTM

**Pull request merged**